### ISSUE TITLE
Error in generating/applying yaml causes helm chart to be redeployed without patchesStrategicMerge values?

### ISSUE THREAD

**Issue by rayisbadat** [fluxcd/flux2#975]
Im sorry if this is confusing. 

Versions
---------
* flux version 0.7.2
* flux components deployed: v0.8.1 (via flux bootstrap)
  * image: ghcr.io/fluxcd/helm-controller:v0.7.0
  * image: ghcr.io/fluxcd/kustomize-controller:v0.8.1
  * image: ghcr.io/fluxcd/notification-controller:v0.8.0
   * image: ghcr.io/fluxcd/source-controller:v0.8.1

Setup
------------
  I am testing out fluxv2 to replace my current fluxv1 instance.  I have my primary git repo, that was deployed with git bootstrap.  I then have a second git repo holding my testing environment.  In that GitRepository i have a "base" directory of the apps themselves, grouping together the GitRepositories and HelmReleases as needed.  I then call specific instances of the apps via subdirectory filled with the relevant `kustomize.toolkit.fluxcd.io`poinitng back to the correct bases and patching the HelmRelease as needed. These kustomizations apply both patchesJson6902 and patchesStrategicMerge patches to set custom settings.  The patchesJson6902 is used to set the namespace, and patchesStrategicMerge for all the values/valuesFrom in the HelmRelease i need to overwrite.  This all works fine at first.

```
├── base
│   └── goHarbor
│       ├── GitRepository.yaml <---source.toolkit.fluxcd.io/v1beta1
│       ├── HelmRelease.yaml <---helm.toolkit.fluxcd.io/v2beta1
│       └── kustomization.yaml  <---kustomize.config.k8s.io/v1beta1
├── fluxv2-dev-apps
│   ├── goHarbor
│   │   ├── goHarbor.yaml <---kustomize.toolkit.fluxcd.io/v1beta1
│   │   └── kustomization.yaml  <---kustomize.config.k8s.io/v1beta1
│   └── kustomization.yaml  <---kustomize.config.k8s.io/v1beta1
└── kustomization.yaml  <---kustomize.config.k8s.io/v1beta1
```
I included a sanitized tarball of the files if it helps understand the above
[dev-cluster.tar.gz](https://github.com/fluxcd/helm-controller/files/6030362/dev-cluster.tar.gz)

Problem
--------
If i do something that cause the rendering of the above yaml to fail.  The relevant Helm chart will still get redeployed/updated  minus the config changes to the HelmRelease that patchesStrategicMerge should be applying.  The patchesJson6902 seem to stay, but that just might be that it is harder for me to mangle and trigger.  I would think that if something errors, the helm chart would NOT be redeployed , rather then redeploy with bad/default values.

I have caused the error two different ways
1.  I introduced some improperly formatted yaml that that would not parse/build properly in the  kustomize.toolkit.fluxcd.io file `goHarbor.yaml`.
  a. Sorry i didn't grab the error msg. But it would be similar to the type of error that would cause`Error: accumulating resources` if trying `kustomize build` on vanilla  k8s kustomize .  
2. I commented out the bases section entirely in kustomize.config.k8s.io  file `fluxv2-dev-apps/kustomization.yaml`.  Which results in fluxv2 (not sure what component) to error, because it generated blank output.   
   a.   ```{"level":"error","ts":"2021-02-23T15:31:26.189Z","logger":"controller.kustomization","msg":"Reconciliation failed after 1.995099153s, next try in 1m0s","reconciler group":"kustomize.toolkit.fluxcd.io","reconciler kind":"Kustomization","name":"dev-team","namespace":"fluxv2-dev-apps","revision":"master/500359cbce24ee4167620e4476b41b9443da1c3b","error":"apply failed: error: no objects passed to apply\n"} ```

Expected Results
------------------
  I think if you get some kind of error building or parsing, it should not update the Helm chart and redeploy it with incorrect values.

**Comment by hiddeco** [comment]
Transferred this issue to `fluxcd/flux2`, as it is not helm-controller related.

My first tip would be to validate the set of YAMLs before committing (or merging) them, there is a hint about how to do this here: https://toolkit.fluxcd.io/faq/#what-is-the-behavior-of-kustomize-used-by-flux

**Comment by rayisbadat** [comment]
A follow up, not sure if its related or separate issue.   

* I deleted the deploy with `flux delete kustomization -n fluxv2-dev-apps goharbor ; flux -n fluxv2-dev-apps delete helmrelease goharbor` 
* I added `targetNamespace: fluxv2-dev-apps-2` to this helmrelease via strategicMergePatch to the kustomization, and ended up with two separate deploys. Both had the correct values in configMaps.  `helm list -A` shows 2 helm charts deployed.
*  I compared `helm get values -n fluxv2-dev-apps fluxv2-dev-apps-goharbor` vs `helm get values -n fluxv2-dev-apps fluxv2-dev-apps-2-goharbor`  and the output was the same  
*  I then manually ran `flux -nfluxv2-dev-apps reconcile helmrelease goharbor` and the instance in `namespace: fluxv2-dev-apps`  updated itself with the unpatched values. 
* I compared `helm get values -n fluxv2-dev-apps fluxv2-dev-apps-goharbor` vs `helm get values -n fluxv2-dev-apps fluxv2-dev-apps-2-goharbor` and now the incorrect one had reverted to the default, unpatched helmrelease
* i did a `flux -nfluxv2-dev-apps delete helmrelease goharbor` and the one in `namespace: fluxv2-dev-apps-2` got deleted as it should.  But the broken one in  `namespace: fluxv2-dev-apps` sticks around and i have to purge manually with a "helm delete".
